atomic_reactor/core.py

```diff
@@ -272,10 +282,57 @@ def __init__(self, base_url=None, retry_times=DOCKER_MAX_RETRIES,
 
         if hasattr(docker, 'AutoVersionClient'):
             client_kwargs['version'] = 'auto'
-        client_kwargs['retry'] = retry_times
+        self.retry_times = retry_times
+        client_kwargs['retry'] = self.retry_times
 
         self.d = WrappedDocker(**client_kwargs)
 
+    def retry_generator(self, function, *args, **kwargs):
+        retry_times = int(kwargs.pop('retry_times', self.retry_times))
+        retry_delay = DOCKER_BACKOFF_FACTOR
+        retry_client_statuses = DOCKER_CLIENT_STATUS_RETRY
+
+        for counter in range(retry_times + 1):
+            exc = None
+            try:
+                logs_gen = function(*args, **kwargs)
+                cmd_result = atomic_reactor.util.wait_for_command(logs_gen)
+
+            except ProtocolError as e:
+                if counter != retry_times:
+                    exc = e
+                else:
+                    raise RetryGeneratorException("Failed to %s image %s: %s" %
+                                                  (function.__name__, args, e.args),
+                                                  e)
+
+            except APIError as e:
+                if (e.response.status_code in retry_client_statuses and counter != retry_times):
+                    exc = e
+                else:
+                    raise RetryGeneratorException("Failed to %s image %s: %r" %
+                                                  (function.__name__, args, e.response.content),
+                                                  e)
+
+            if exc:
+                if (isinstance(exc, ProtocolError)):
```

Comment:
Does it make sense to move these log messages to their corresponding except-block?
This seems error prone (what if start catching another exception?)